src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -289,6 +304,255 @@ public struct S
+                //     public readonly override bool Equals(object o) => (i++).Equals(o);
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(9, 56));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_GetPinnableReference()
+        {
+            var csharp = @"
+public unsafe struct S1
+{
+    static int i = 0;
+    ref int GetPinnableReference() => ref i;
+    void M1()
+    {
+        fixed (int *i = this) {} // ok
+    }
+    readonly void M2()
+    {
+        fixed (int *i = this) {} // warn
+    }
+}
+
+public unsafe struct S2
+{
+    static int i = 0;
+    readonly ref int GetPinnableReference() => ref i;
+    void M1()
+    {
+        fixed (int *i = this) {} // ok
+    }
+    readonly void M2()
+    {
+        fixed (int *i = this) {} // ok
+    }
+}
+";
+            var comp = CreateCompilation(csharp, options: TestOptions.UnsafeReleaseDll);
+            comp.VerifyDiagnostics(
+                // (12,25): warning CS8655: Call to non-readonly member 'GetPinnableReference' from a 'readonly' member results in an implicit copy of 'this'.
+                //         fixed (int *i = this) {} // warn
+                Diagnostic(ErrorCode.WRN_ImplicitCopyInReadOnlyMember, "this").WithArguments("GetPinnableReference", "this").WithLocation(12, 25));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_Iterator()
+        {
+            var csharp = @"
+using System.Collections.Generic;
+
+public struct S
+{
+    public int i;
+    public readonly IEnumerable<int> M1()
+    {
+        yield return i;
+        yield return i+1;
+    }
+
+    public readonly IEnumerable<int> M2()
+    {
+        yield return i;
+        i++;
+        yield return i;
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (16,9): error CS1604: Cannot assign to 'this' because it is read-only
+                //         i++;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(16, 9));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_Async()
+        {
+            var csharp = @"
+using System.Threading.Tasks;
+
+public struct S
+{
+    public int i;
+    public readonly async Task<int> M1()
+    {
+        await Task.Delay(1);
+        return i;
+    }
+
+    public readonly async Task<int> M2()
+    {
+        await Task.Delay(1);
+        i++;
+        await Task.Delay(1);
+        return i;
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (16,9): error CS1604: Cannot assign to 'this' because it is read-only
+                //         i++;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(16, 9));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_AsyncStreams()
```

Comment:
Nit: consider putting this test with the `GetEnumerator` and family tests, as they're all tests of various `foreach` things.

Reply:
Done